system/include/emscripten/em_asm.h

```diff
@@ -1,35 +1,84 @@
 #ifndef __em_asm_h__
 #define __em_asm_h__
 
-// The wasm backend calls vararg functions by passing the variadic arguments as
-// an array on the stack. For EM_ASM's underlying functions, s2wasm needs to
-// translate them to emscripten_asm_const_[signature], but the expected
-// signature and arguments has been lost as part of the vararg buffer.
-// Therefore, we declare EM_ASM's implementing functions as non-variadic.
+#ifdef __cplusplus
+extern "C" {
+#endif // __cplusplus
 
 #ifndef __asmjs
-#ifndef __cplusplus
-// In C, declare these as non-prototype declarations. This is obsolete K&R C
-// (that is still supported) that causes the C frontend to consider any calls
-// to them as valid, and avoids using the vararg calling convention.
-void emscripten_asm_const();
-int emscripten_asm_const_int();
-double emscripten_asm_const_double();
-#else
-// C++ interprets an empty parameter list as a function taking no arguments,
-// instead of a K&R C function declaration. Variadic templates are lowered as
-// non-vararg calls to the instantiated templated function, which we then
-// replace in s2wasm.
-template <typename... Args> void emscripten_asm_const(const char* code, Args...);
-template <typename... Args> int emscripten_asm_const_int(const char* code, Args...);
-template <typename... Args> double emscripten_asm_const_double(const char* code, Args...);
-#endif // __cplusplus
+// When calling EM_ASM functions, we're calling out to JS. JS only has doubles
+// natively, so we can convert all C types to double before the call. By
+// converting all arguments to double, we can store them in a vararg buffer,
+// and we only need to pass the number of args in order to read them in JS.
+
+// When converting arguments to double, we need to:
+//   - upcast floats to doubles, so we have consistent bit widths
+//   - convert ints to doubles (via f64.convert_s/i32 and such)
+//   - cast pointers to ints, and then convert (C disallows direct pointer to
+//     double conversions)
+// We can use the _Generic clang extension / C11 feature to decide at compile
+// time what operations to apply. Unfortunately each branch needs to compile for
+// all possible types, even though only one is selected. Fortunately, we can use
+// a function inside the _Generic expression, and use that do do our conversion.
+// All pointer types should go through the default: case.
+#define _EA_CAST(x) _Generic((x), \
+    float: _ea_cast_double, \
+    double: _ea_cast_double, \
+    int: _ea_cast_int, \
+    unsigned: _ea_cast_uint, \
+    long: _ea_cast_int, \
+    unsigned long: _ea_cast_uint, \
+    default: _ea_cast_ptr)(x)
+inline double _ea_cast_double(double x) {
+  return x;
+}
+inline double _ea_cast_int(long x) {
+  return (double)x;
+}
+inline double _ea_cast_uint(unsigned long x) {
+  return (double)x;
+}
+inline double _ea_cast_ptr(const void* x) {
+  return (double)(long)x;
```

Comment:
could `long` be `size_t`?

Reply:
`size_t` turns out to not be defined by that point. I think in principle it should be that, but `long` should conservatively work too?

Reply:
I guess so, yeah.